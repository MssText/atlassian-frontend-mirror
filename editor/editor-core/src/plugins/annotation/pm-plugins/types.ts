import { PortalProviderAPI } from '../../../ui/PortalProvider';
import { Dispatch, EventDispatcher } from '../../../event-dispatcher';
import { AnnotationTypeProviders } from '../types';

export interface InlineCommentPluginOptions {
  dispatch: Dispatch;
  eventDispatcher: EventDispatcher;
  portalProviderAPI: PortalProviderAPI;
  inlineCommentProvider: AnnotationTypeProviders['inlineComment'];
  pollingInterval?: number;
}
