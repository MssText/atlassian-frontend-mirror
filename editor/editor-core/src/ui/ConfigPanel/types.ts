import {
  ExtensionProvider,
  ExtensionType,
  ExtensionKey,
  Parameters,
} from '@atlaskit/editor-common/extensions';

export type FormResult = {
  [key: string]: string | number | string[] | number[] | undefined;
};

export enum ValidationError {
  Required = 'required',
}

export enum FieldTypeError {
  isMultipleAndRadio = 'isMultipleAndRadio',
}

export type Entry<T> = [string, T];

export type PublicProps = {
  extensionProvider: ExtensionProvider;
  extensionType: ExtensionType;
  extensionKey: ExtensionKey;
  nodeKey: string;
  extensionParameters?: Parameters;
  parameters?: Parameters;
  autoSave?: boolean;
  closeOnEsc?: boolean;
  showHeader?: boolean;
  onChange: (data: Parameters) => void;
  onCancel: () => void;
};

export type OnBlur = (name: string) => void;
