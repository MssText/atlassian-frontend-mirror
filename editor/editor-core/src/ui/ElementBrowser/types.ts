export type CategoryType = {
  title: string;
  name: string;
  selected: boolean;
};

export enum Modes {
  full = 'full',
  inline = 'inline',
}
