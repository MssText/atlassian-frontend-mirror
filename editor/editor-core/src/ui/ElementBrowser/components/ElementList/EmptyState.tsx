import React from 'react';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';

import OpenIcon from '@atlaskit/icon/glyph/open';
import NotFoundIllustration from './NotFoundIllustration';

type Props = {
  onExternalLinkClick: () => void;
};

export default function EmptyState({
  onExternalLinkClick,
}: Props): JSX.Element {
  return (
    <EmptyStateWrapper>
      <NotFoundIllustration />
      <EmptyStateHeading>
        <FormattedMessage
          id="fabric.editor.elementbrowser.search.empty-state.heading"
          defaultMessage="Nothing matches your search"
          description="Empty state heading"
        />
      </EmptyStateHeading>
      <EmptyStateSubHeading>
        <FormattedMessage
          id="fabric.editor.elementbrowser.search.empty-state.sub-heading"
          defaultMessage="Try searching again using different terms or"
          description="Empty state sub-heading"
        />
        <br />
        <ExternalLink
          target="_blank"
          href="https://marketplace.atlassian.com/search?category=Macros&hosting=cloud&product=confluence"
          onClick={onExternalLinkClick}
        >
          <FormattedMessage
            id="fabric.editor.elementbrowser.search.empty-state.sub-heading.link"
            defaultMessage="explore apps from Atlassian Marketplace"
            description="Empty state sub-heading external link"
          />
          <OpenIcon label="external-link" />
        </ExternalLink>
      </EmptyStateSubHeading>
    </EmptyStateWrapper>
  );
}

const EmptyStateHeading = styled.div`
  font-size: 1.42857em;
  line-height: 1.2;
  color: rgb(23, 43, 77);
  font-weight: 500;
  letter-spacing: -0.008em;
  margin-top: 28px;
`;

const EmptyStateSubHeading = styled.p`
  max-width: 400px;
  text-align: center;
`;

const EmptyStateWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
`;

const ExternalLink = styled.a`
  display: inline-flex;
  align-items: center;
`;
