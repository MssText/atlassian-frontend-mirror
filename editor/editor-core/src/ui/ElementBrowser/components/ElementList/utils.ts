import { FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS } from '../../constants';

export function getColumnCount(clientWidth: number): number {
  const { small, medium, large } = FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS;
  switch (true) {
    case clientWidth < small:
      return 1;
    case clientWidth >= small && clientWidth < medium:
      return Math.floor(clientWidth / 200);
    case clientWidth >= large:
      return Math.floor(clientWidth / 248);
    default:
      return Math.floor(clientWidth / 220);
  }
}

type DatumReturnType = {
  availableWidth: number;
  columnCount: number;
};

type Options = {
  gutterSize: number;
  scrollbarWidth: number;
};

export function generateVirtualizedContainerDatum(
  containerWidth: number,
  options: Options,
): DatumReturnType {
  const { gutterSize, scrollbarWidth } = options;
  const columnCount = getColumnCount(containerWidth);
  const extraSpace = columnCount * gutterSize;
  const innerContainerWidth = containerWidth - (scrollbarWidth + gutterSize);
  const availableWidth = innerContainerWidth - extraSpace;
  return {
    availableWidth,
    columnCount,
  };
}
