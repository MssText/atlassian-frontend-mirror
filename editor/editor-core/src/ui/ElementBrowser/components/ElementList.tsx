import React, { memo, useCallback, useEffect, useMemo } from 'react';
import { FormattedMessage } from 'react-intl';
import memoizeOne from 'memoize-one';
import styled, { css, ThemeProvider } from 'styled-components';
import Tooltip from '@atlaskit/tooltip';
import { QuickInsertItem } from '@atlaskit/editor-common/provider-factory';
import { N200 } from '@atlaskit/theme/colors';
import Item from '@atlaskit/item';
import { withAnalyticsContext } from '@atlaskit/analytics-next';
import { Shortcut } from '../../styles';
import { ItemIcon } from '../../../plugins/type-ahead/ui/TypeAheadItemsList';
import IconFallback from '../../../plugins/quick-insert/assets/fallback';
import {
  GRID_SIZE,
  SCROLLBAR_THUMB_COLOR,
  SCROLLBAR_TRACK_COLOR,
  FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS,
} from '../constants';
import useRefToFocusOrScroll from '../hooks/useRefToFocusOrScroll';
import { Modes, SelectedItemProps } from '../types';
import useContainerWidth from '../hooks/useContainerWidth';
import EmptyState from './EmptyState';

interface Props {
  items: QuickInsertItem[];
  mode: keyof typeof Modes;
  onSelectItem: (item: QuickInsertItem) => void;
  setItemsContainerWidth?: (width: number) => void;
  setFocusedItemIndex: (index: number) => void;
}

function ElementList({
  items,
  mode,
  selectedItemIndex,
  focusedItemIndex,
  setItemsContainerWidth,
  ...props
}: Props & SelectedItemProps) {
  const { containerWidth, ContainerWidthMonitor } = useContainerWidth();

  useEffect(() => {
    if (setItemsContainerWidth) {
      setItemsContainerWidth(containerWidth);
    }
  }, [containerWidth, setItemsContainerWidth]);

  const theme = useMemo(
    () => ({
      '@atlaskit-shared-theme/item': getStyles(mode, containerWidth),
    }),
    [mode, containerWidth],
  );
  return (
    <>
      <ContainerWidthMonitor />
      <ElementItemsWrapper mode={mode} tabIndex={-1}>
        <ThemeProvider theme={theme}>
          <>
            {!items.length ? (
              <EmptyStateWrapper>
                <EmptyState />
                <EmptyStateHeading>
                  <FormattedMessage
                    id="fabric.editor.elementbrowser.search.empty-state.heading"
                    defaultMessage="Nothing matches your search"
                    description="Empty state heading"
                  />
                </EmptyStateHeading>
                <EmptyStateSubHeading>
                  <FormattedMessage
                    id="fabric.editor.elementbrowser.search.empty-state.sub-heading"
                    defaultMessage="Remove any filters, or search for something less specific."
                    description="Empty state sub-heading"
                  />
                </EmptyStateSubHeading>
              </EmptyStateWrapper>
            ) : (
              items.map<JSX.Element>(
                (item, index): JSX.Element => (
                  <MemoizedElementItem
                    key={`${item.title}-${item.description}`}
                    index={index}
                    item={item}
                    selected={selectedItemIndex === index}
                    focus={focusedItemIndex === index}
                    {...props}
                  />
                ),
              )
            )}
          </>
        </ThemeProvider>
      </ElementItemsWrapper>
    </>
  );
}

const getStyles = memoizeOne((mode, containerWidth) => {
  return {
    ...(mode === Modes.full && {
      width: {
        default: getItemWidthForContainer(containerWidth),
      },
      '-ms-flex': 'auto',
      position: 'relative',
      boxSizing: 'border-box',
    }),
    height: {
      default: GRID_SIZE * GRID_SIZE,
    },
    padding: {
      default: {
        top: GRID_SIZE * 1.5,
        right: GRID_SIZE * 1.5,
        bottom: GRID_SIZE * 1.5,
        left: GRID_SIZE * 1.5,
      },
    },
    borderRadius: GRID_SIZE / 2,
    hover: {
      background: 'rgb(244, 245, 247)',
    },
    beforeItemSpacing: {
      default: GRID_SIZE * 1.5,
    },
  };
});

const getItemWidthForContainer = (clientWidth?: number) => {
  if (clientWidth == null) {
    return '33%';
  }
  if (clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.small) {
    return '100%';
  }
  if (
    clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.small &&
    clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.medium
  ) {
    return '50%';
  }
  if (
    clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.medium &&
    clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.large
  ) {
    return '33%';
  }
  if (clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.large) {
    return '25%';
  }
};

export const itemStepCounter = (clientWidth: number): number => {
  if (clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.small) {
    return 1;
  }
  if (
    clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.small &&
    clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.medium
  ) {
    return 2;
  }
  if (
    clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.medium &&
    clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.large
  ) {
    return 3;
  }
  if (clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.large) {
    return 4;
  }
  return 1;
};

type ElementItemType = {
  item: QuickInsertItem;
  onSelectItem: (item: QuickInsertItem) => void;
  selected: boolean;
  focus: boolean;
  setFocusedItemIndex: (index: number) => void;
  index: number;
};

const MemoizedElementItem = memo(ElementItem);
MemoizedElementItem.displayName = 'MemoizedElementItem';

function ElementItem({
  selected,
  item,
  index,
  onSelectItem,
  focus,
  setFocusedItemIndex,
}: ElementItemType) {
  const ref = useRefToFocusOrScroll(focus, selected);

  const onClick = useCallback(
    (e: React.MouseEvent) => {
      e.preventDefault();
      e.stopPropagation();

      setFocusedItemIndex(index);
      onSelectItem(item);
    },
    [index, item, onSelectItem, setFocusedItemIndex],
  );

  // After tabbing we wanna select the item on enter/space key press from item level,
  // preventing the default top level component behavior.
  const onKeyPress = useCallback(
    (e: React.KeyboardEvent) => {
      const SPACE_KEY = 32;
      const ENTER_KEY = 13;

      if (e.which === ENTER_KEY || e.which === SPACE_KEY) {
        e.preventDefault();
        e.stopPropagation();
        setFocusedItemIndex(index);
        onSelectItem(item);
      }
    },
    [index, item, onSelectItem, setFocusedItemIndex],
  );

  return (
    <Item
      onClick={onClick}
      elemBefore={
        <StyledItemIcon>
          {item.icon ? item.icon() : <IconFallback label={item.title} />}
        </StyledItemIcon>
      }
      isSelected={selected}
      aria-describedby={item.title}
      innerRef={ref}
      onKeyPress={onKeyPress}
    >
      <Tooltip content={item.description}>
        <ItemBody>
          <ItemText>
            <div>{item.title}</div>
            {item.description && (
              <ItemDescription>{item.description}</ItemDescription>
            )}
          </ItemText>
          <ItemAfter>
            {item.keyshortcut && <Shortcut>{item.keyshortcut}</Shortcut>}
          </ItemAfter>
        </ItemBody>
      </Tooltip>
    </Item>
  );
}

interface ContainerProps {
  mode: keyof typeof Modes;
}

const scrollbarStyle = css`
  ::-webkit-scrollbar {
    width: ${GRID_SIZE}px;
  }
  ::-webkit-scrollbar-track-piece {
    background: ${SCROLLBAR_TRACK_COLOR};
  }
  ::-webkit-scrollbar-thumb {
    background: ${SCROLLBAR_THUMB_COLOR};
  }

  /** Firefox **/
  scrollbar-color: ${SCROLLBAR_THUMB_COLOR} ${SCROLLBAR_TRACK_COLOR};

  -ms-overflow-style: -ms-autohiding-scrollbar;
`;

const ElementItemsWrapper = styled.div`
  display: ${({ mode }: ContainerProps) =>
    mode === Modes.full ? 'flex' : 'block'};
  flex-flow: row wrap;
  align-items: flex-start;
  justify-content: flex-start;
  overflow-x: hidden;
  overflow-y: scroll;
  ${scrollbarStyle};
`;

const ItemBody = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
  line-height: 1.4;
  width: 100%;
`;

const ItemDescription = styled.div`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 11.67px;
  color: ${N200};
  margin-top: ${GRID_SIZE / 2}px;
`;

const ItemText = styled.div`
  width: inherit;
  white-space: initial;
`;

const ItemAfter = styled.div`
  flex: 0 0 auto;
`;

const StyledItemIcon = styled(ItemIcon)`
  img {
    height: 40px;
    width: 40px;
    object-fit: cover;
  }
`;

const EmptyStateHeading = styled.div`
  font-size: 1.42857em;
  line-height: 1.2;
  color: rgb(23, 43, 77);
  font-weight: 500;
  letter-spacing: -0.008em;
  margin-top: 28px;
`;

const EmptyStateSubHeading = styled.p`
  max-width: 400px;
  text-align: center;
`;

const EmptyStateWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100%;
  width: 100%;
`;

const MemoizedElementListWithAnalytics = memo(
  withAnalyticsContext({ component: 'ElementList' })(ElementList),
);

export default MemoizedElementListWithAnalytics;
