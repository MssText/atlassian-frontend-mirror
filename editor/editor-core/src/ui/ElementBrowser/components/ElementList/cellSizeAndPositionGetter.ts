import {
  CollectionCellSizeAndPosition,
  CollectionCellSizeAndPositionGetter,
} from 'react-virtualized/dist/commonjs/Collection';
import { GRID_SIZE, SCROLLBAR_WIDTH } from '../../constants';
import { generateVirtualizedContainerDatum } from './utils';

/**
 * Callback responsible for returning size and offset/position information
 * for a given cell.
 * https://github.com/bvaughn/react-virtualized/blob/master/docs/Collection.md
 **/
export default function cellSizeAndPositionGetter(
  containerWidth: number,
): CollectionCellSizeAndPositionGetter {
  const GUTTER_SIZE = 4;
  /**
   * Save the currently rendered columnY positions.
   * Have to be within the current render scope.
   */
  let columnYMap: number[] = [];
  return ({ index }): CollectionCellSizeAndPosition => {
    const { columnCount, availableWidth } = generateVirtualizedContainerDatum(
      containerWidth,
      {
        gutterSize: GUTTER_SIZE,
        scrollbarWidth: SCROLLBAR_WIDTH,
      },
    );

    const width = Math.floor(availableWidth / columnCount);

    const height = GRID_SIZE * GRID_SIZE;

    const columnPosition = index % (columnCount || 1);

    const x = columnPosition * (GUTTER_SIZE + width);
    const y = columnYMap[columnPosition] || 0;

    columnYMap[columnPosition] = y + height + GUTTER_SIZE;

    return {
      height,
      width,
      x,
      y,
    };
  };
}
