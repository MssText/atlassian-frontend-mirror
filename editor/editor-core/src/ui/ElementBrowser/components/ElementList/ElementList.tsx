import React, { memo, useCallback, useEffect, useMemo } from 'react';
import memoizeOne from 'memoize-one';
import styled, { ThemeProvider, css } from 'styled-components';
import { AutoSizer, Size } from 'react-virtualized/dist/commonjs/AutoSizer';
import { Collection } from 'react-virtualized/dist/commonjs/Collection';
import { QuickInsertItem } from '@atlaskit/editor-common/provider-factory';
import Item from '@atlaskit/item';
import { N20, N200 } from '@atlaskit/theme/colors';
import Tooltip from '@atlaskit/tooltip';

import {
  withAnalyticsContext,
  WithAnalyticsEventsProps,
} from '@atlaskit/analytics-next';
import {
  ACTION,
  ACTION_SUBJECT,
  EVENT_TYPE,
  fireAnalyticsEvent,
} from '../../../../plugins/analytics';

import IconFallback from '../../../../plugins/quick-insert/assets/fallback';
import { ItemIcon } from '../../../../plugins/type-ahead/ui/TypeAheadItemsList';
import { Shortcut } from '../../../styles';
import {
  GRID_SIZE,
  SCROLLBAR_THUMB_COLOR,
  SCROLLBAR_TRACK_COLOR,
  SCROLLBAR_WIDTH,
} from '../../constants';
import useContainerWidth from '../../hooks/use-container-width';
import useFocus from '../../hooks/use-focus';
import { Modes, SelectedItemProps } from '../../types';
import cellSizeAndPositionGetter from './cellSizeAndPositionGetter';
import EmptyState from './EmptyState';
import { getColumnCount } from './utils';

export interface Props {
  items: QuickInsertItem[];
  mode: keyof typeof Modes;
  onInsertItem: (item: QuickInsertItem) => void;
  setColumnCount: (columnCount: number) => void;
  setFocusedItemIndex: (index: number) => void;
}

function ElementList({
  items,
  mode,
  selectedItemIndex,
  focusedItemIndex,
  setColumnCount,
  createAnalyticsEvent,
  ...props
}: Props & SelectedItemProps & WithAnalyticsEventsProps) {
  const { containerWidth, ContainerWidthMonitor } = useContainerWidth();

  const fullMode = mode === Modes.full;

  useEffect(() => {
    /**
     * More of an optimization condition.
     * Initially the containerWidths are reported 0 twice.
     **/
    if (fullMode && containerWidth > 0) {
      setColumnCount(getColumnCount(containerWidth));
    }
    removeCollectionTabIndex();
  }, [fullMode, containerWidth, setColumnCount]);

  const onExternalLinkClick = useCallback(() => {
    fireAnalyticsEvent(createAnalyticsEvent)({
      payload: {
        action: ACTION.VISITED,
        actionSubject: ACTION_SUBJECT.SMART_LINK,
        eventType: EVENT_TYPE.TRACK,
      },
    });
  }, [createAnalyticsEvent]);

  const theme = useMemo(
    () => ({
      '@atlaskit-shared-theme/item': getStyles(mode),
    }),
    [mode],
  );
  const cellRenderer = useMemo(
    () => ({
      index,
      key,
      style,
    }: {
      index: number;
      key: string | number;
      style: object;
    }) => {
      if (items[index] == null) {
        return;
      }

      return (
        <div style={style} key={key} className="element-item-wrapper">
          <MemoizedElementItem
            inlineMode={!fullMode}
            index={index}
            item={items[index]}
            selected={selectedItemIndex === index}
            focus={focusedItemIndex === index}
            {...props}
          />
        </div>
      );
    },
    [items, fullMode, selectedItemIndex, focusedItemIndex, props],
  );
  return (
    <>
      <ContainerWidthMonitor />
      {!items.length ? (
        <EmptyState onExternalLinkClick={onExternalLinkClick} />
      ) : (
        <ElementItemsWrapper tabIndex={-1} data-testid="ElementItems">
          <ThemeProvider theme={theme}>
            <>
              {containerWidth > 0 && (
                <AutoSizer disableWidth>
                  {({ height }: Size) => (
                    <Collection
                      cellCount={items.length}
                      cellRenderer={cellRenderer}
                      cellSizeAndPositionGetter={cellSizeAndPositionGetter(
                        containerWidth,
                      )}
                      height={height}
                      width={containerWidth}
                      key={containerWidth} // Refresh Collection on WidthObserver value change.
                      scrollToCell={selectedItemIndex}
                    />
                  )}
                </AutoSizer>
              )}
            </>
          </ThemeProvider>
        </ElementItemsWrapper>
      )}
    </>
  );
}

const getStyles = memoizeOne(mode => {
  return {
    ...(mode === Modes.full && {
      '-ms-flex': 'auto',
      position: 'relative',
      boxSizing: 'border-box',
    }),
    height: {
      default: GRID_SIZE * GRID_SIZE,
    },
    padding: {
      default: {
        top: GRID_SIZE * 1.5,
        right: GRID_SIZE * 1.5,
        bottom: GRID_SIZE * 1.5,
        left: GRID_SIZE * 1.5,
      },
    },
    borderRadius: GRID_SIZE / 2,
    selected: {
      background: N20,
    },
    hover: {
      background: 'rgb(244, 245, 247)',
    },
    beforeItemSpacing: {
      default: GRID_SIZE * 1.5,
    },
  };
});

type ElementItemType = {
  inlineMode: boolean;
  item: QuickInsertItem;
  onInsertItem: (item: QuickInsertItem) => void;
  selected: boolean;
  focus: boolean;
  setFocusedItemIndex: (index: number) => void;
  index: number;
};

const MemoizedElementItem = memo(ElementItem);
MemoizedElementItem.displayName = 'MemoizedElementItem';

function ElementItem({
  inlineMode,
  selected,
  item,
  index,
  onInsertItem,
  focus,
  setFocusedItemIndex,
}: ElementItemType) {
  const ref = useFocus(focus);

  /**
   * Note: props.onSelectItem(item) is not called here as the StatelessElementBrowser's
   * useEffect would trigger it on selectedItemIndex change. (Line 106-110)
   * This implementation was changed for keyboard/click selection to work with `onInsertItem`.
   */
  const onClick = useCallback(
    (e: React.MouseEvent) => {
      e.preventDefault();
      e.stopPropagation();
      setFocusedItemIndex(index);
      if (inlineMode) {
        onInsertItem(item);
      }
    },
    [index, inlineMode, item, onInsertItem, setFocusedItemIndex],
  );

  const onDoubleClick = useCallback(
    (e: React.MouseEvent) => {
      if (inlineMode) {
        return;
      }
      e.preventDefault();
      e.stopPropagation();
      setFocusedItemIndex(index);
      onInsertItem(item);
    },
    [inlineMode, setFocusedItemIndex, index, onInsertItem, item],
  );

  // After tabbing we wanna select the item on enter/space key press from item level,
  // preventing the default top level component behavior.
  const onKeyPress = useCallback(
    (e: React.KeyboardEvent) => {
      const SPACE_KEY = 32;
      const ENTER_KEY = 13;

      if (e.which === ENTER_KEY || e.which === SPACE_KEY) {
        e.preventDefault();
        e.stopPropagation();
        setFocusedItemIndex(index);
        onInsertItem(item);
      }
    },
    [index, item, onInsertItem, setFocusedItemIndex],
  );

  const { icon, title, description, keyshortcut } = item;
  return (
    <Item
      onClick={onClick}
      onDoubleClick={onDoubleClick}
      elemBefore={<ElementBefore icon={icon} title={title} />}
      isSelected={selected}
      aria-describedby={title}
      innerRef={ref}
      onKeyPress={onKeyPress}
      data-testid={`element-item-${index}`}
    >
      <ItemContent
        title={title}
        description={description}
        keyshortcut={keyshortcut}
      />
    </Item>
  );
}

const ElementBefore = memo(({ icon, title }: Partial<QuickInsertItem>) => (
  <StyledItemIcon>
    {icon ? icon() : <IconFallback label={title} />}
  </StyledItemIcon>
));

const ItemContent = memo(
  ({ title, description, keyshortcut }: Partial<QuickInsertItem>) => (
    <Tooltip content={description}>
      <ItemBody>
        <ItemText>
          <div>{title}</div>
          {description && <ItemDescription>{description}</ItemDescription>}
        </ItemText>
        <ItemAfter>
          {keyshortcut && <Shortcut>{keyshortcut}</Shortcut>}
        </ItemAfter>
      </ItemBody>
    </Tooltip>
  ),
);

const scrollbarStyle = css`
  ::-webkit-scrollbar {
    width: ${SCROLLBAR_WIDTH}px;
  }
  ::-webkit-scrollbar-track-piece {
    background: ${SCROLLBAR_TRACK_COLOR};
  }
  ::-webkit-scrollbar-thumb {
    background: ${SCROLLBAR_THUMB_COLOR};
  }

  /** Firefox **/
  scrollbar-color: ${SCROLLBAR_THUMB_COLOR} ${SCROLLBAR_TRACK_COLOR};

  -ms-overflow-style: -ms-autohiding-scrollbar;
`;

const ElementItemsWrapper = styled.div`
  flex: 1;
  flex-flow: row wrap;
  align-items: flex-start;
  justify-content: flex-start;
  overflow: hidden;

  .ReactVirtualized__Collection {
    ${scrollbarStyle};
  }
  .ReactVirtualized__Collection__innerScrollContainer {
    div[class='element-item-wrapper']:last-child {
      padding-bottom: 4px;
    }
  }
`;

const ItemBody = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
  line-height: 1.4;
  width: 100%;
`;

const ItemDescription = styled.div`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 11.67px;
  color: ${N200};
  margin-top: ${GRID_SIZE / 2}px;
`;

const ItemText = styled.div`
  width: inherit;
  white-space: initial;
`;

const ItemAfter = styled.div`
  flex: 0 0 auto;
`;

const StyledItemIcon = styled(ItemIcon)`
  img {
    height: 40px;
    width: 40px;
    object-fit: cover;
  }
`;

/**
 *
 * Internally, the Collection component has a tabIndex of 0 and we don't wanna focus on the entire Collection area,
 * so manually removing it until the below PR has been approved/merged.
 * https://product-fabric.atlassian.net/browse/ED-9919
 * https://github.com/bvaughn/react-virtualized/pull/1555
 */
const removeCollectionTabIndex = () => {
  const element = document.getElementsByClassName(
    'ReactVirtualized__Collection',
  )[0] as HTMLElement;
  if (element && element.tabIndex !== -1) {
    element.tabIndex = -1;
  }
};

const MemoizedElementListWithAnalytics = memo(
  withAnalyticsContext({ component: 'ElementList' })(ElementList),
);

export default MemoizedElementListWithAnalytics;
