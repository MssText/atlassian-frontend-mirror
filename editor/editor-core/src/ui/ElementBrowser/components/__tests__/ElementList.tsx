const getWidth = jest.fn();

jest.mock('../../hooks/useContainerWidth', () => {
  const originalHook = jest.requireActual('../../hooks/useContainerWidth')
    .default;
  return (...args: any) => ({
    ...originalHook(...args),
    containerWidth: getWidth(),
  });
});

import React from 'react';
import { shallow, ReactWrapper, mount } from 'enzyme';
import ElementList from '../ElementList';
import { Modes } from '../../types';

const props = {
  items: [
    { name: 'item-1', title: 'Item 1', action: jest.fn() },
    { name: 'item-2', title: 'Item 2', action: jest.fn() },
    { name: 'item-3', title: 'Item 3', action: jest.fn() },
    { name: 'item-4', title: 'Item 4', action: jest.fn() },
    { name: 'item-5', title: 'Item 5', action: jest.fn() },
  ],
  mode: Modes.full,
  onSelectItem: jest.fn(),
  onEnterKeyPress: jest.fn(),
  selectedItemIndex: 1,
  setSelectedItemIndex: jest.fn(),
  setFocusedItemIndex: jest.fn(),
  setItemsContainerWidth: jest.fn(),
};

const getThemeProviderWidth = (wrapper: any) => {
  return wrapper.find('ThemeProvider').prop('theme')[
    '@atlaskit-shared-theme/item'
  ].width.default;
};

let wrapper: ReactWrapper;

describe('ElementList', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<ElementList {...props} />);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });
  describe('getItemWidthForContainer', () => {
    it('should get 50% width for an item if the container width is >= 400px', () => {
      getWidth.mockReturnValueOnce(500);
      wrapper = mount(<ElementList {...props} />);
      expect(getThemeProviderWidth(wrapper)).toStrictEqual('50%');
      wrapper.unmount();
    });
    it('should get 33% width for an item if the container width is >= 600px and < 1024px', () => {
      getWidth.mockReturnValueOnce(601);
      wrapper = mount(<ElementList {...props} />);
      expect(getThemeProviderWidth(wrapper)).toStrictEqual('33%');
      wrapper.unmount();
    });
    it('should get 25% width for an item if the container width is >= 1024px', () => {
      getWidth.mockReturnValueOnce(1025);
      wrapper = mount(<ElementList {...props} />);
      expect(getThemeProviderWidth(wrapper)).toStrictEqual('25%');
      wrapper.unmount();
    });
  });
  it('should render ElementItem component', () => {
    const wrapper = mount(<ElementList {...props} />);
    expect(wrapper.find('MemoizedElementItem').first()).toHaveLength(1);
    wrapper.unmount();
  });
});
