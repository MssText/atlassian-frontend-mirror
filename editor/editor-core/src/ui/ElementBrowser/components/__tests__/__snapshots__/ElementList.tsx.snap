// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementList should match snapshot 1`] = `
<AnalyticsContext
  data={
    Object {
      "component": "ElementList",
    }
  }
>
  <ElementList
    items={
      Array [
        Object {
          "action": [MockFunction],
          "name": "item-1",
          "title": "Item 1",
        },
        Object {
          "action": [MockFunction],
          "name": "item-2",
          "title": "Item 2",
        },
        Object {
          "action": [MockFunction],
          "name": "item-3",
          "title": "Item 3",
        },
        Object {
          "action": [MockFunction],
          "name": "item-4",
          "title": "Item 4",
        },
        Object {
          "action": [MockFunction],
          "name": "item-5",
          "title": "Item 5",
        },
      ]
    }
    mode="full"
    onEnterKeyPress={[MockFunction]}
    onSelectItem={[MockFunction]}
    selectedItemIndex={1}
    setFocusedItemIndex={[MockFunction]}
    setItemsContainerWidth={[MockFunction]}
    setSelectedItemIndex={[MockFunction]}
  />
</AnalyticsContext>
`;
